### PR TITLE
Integration branch (Uniform refine earlier)

### PR DESCRIPTION
**Pull request by permcody**
This PR moves the uniform refinement steps up into the mesh setup stage. The original code path remains for the case where restart and uniform refinements are both needed. 

Important note: The definition of what uniform_refinement means has changed for oversampled meshes. Previously that number assumed how many refinements you wanted from the original mesh, now we have defined it to mean the number of refinements to apply after the mesh has been setup which included the initial uniform refinement steps.

Includes a small bug fix for setting _current_task when executing Actions individually.

### CODE REVIEW

File: framework/src/actions/ActionWarehouse.C [hunk 1]
```diff
@@ -312,6 +309,9 @@ ActionWarehouse::executeAllActions()
 void
 ActionWarehouse::executeActionsWithAction(const std::string & task)
 {
+  // Set the current task name
+  _current_task = task;
```
Comment (friedmud) [inline framework/src/actions/ActionWarehouse.C]: This is a good change.  Was it actually required to make all of this work?

Note: I'm not criticizing, just curious...

Comment (permcody) [inline framework/src/actions/ActionWarehouse.C]: Well it was in order to clean up MooseApp.C where I added the new "uniform_refine_mesh" and had it fire an existing Action manually. There's no place in the code where we hitting this potential bug yet but now it's fixed.

File: framework/src/actions/SetupMeshCompleteAction.C [hunk 2]
```diff
@@ -49,8 +51,24 @@ SetupMeshCompleteAction::act()
...
     mooseError("No mesh file was supplied and no generation block was provided");
 
-  completeSetup(_mesh.get());
+  /**
+   * If possible we'd like to refine the mesh here before the equation systems
+   * are setup to avoid doing expensive projections. If however we are doing a
+   * file based restart and we need uniform refinements, we'll have to postpone
+   * those refinements until after the solution has been read in.
+   */
+  if (_current_task == "uniform_refine_mesh" && _app.setFileRestart() == false)
```
Comment (friedmud) [inline framework/src/actions/SetupMeshCompleteAction.C]: Cool: this will save a ton of time on bigger problems (especially on lots of processors)

File: framework/src/actions/ActionWarehouse.C (hunk 1 above)
Comment (friedmud) [inline framework/src/actions/ActionWarehouse.C]: Cool: that was definitely a bug waiting to happen

File: test/tests/outputs/oversample/oversample_file.i [hunk 3]
```diff
@@ -42,7 +42,7 @@
 [Outputs]
   [./exodus]
     type = Exodus
-    refinements = 2
+    refinements = 1
```
Comment (friedmud) [inline test/tests/outputs/oversample/oversample_file.i]: I understand this better now... And this way actually makes more sense I think.  Closer to what a new user would expect.

Comment (permcody) [inline test/tests/outputs/oversample/oversample_file.i]: It affects about 10 tests in Marmot and that's it. We talked it over and all think it makes more sense here as well.

**Comment by permcody** [comment]
Well I'm currently stumped. This PR simply does not work with the stateful adaptivity system in parallel and I don't fully understand why. When running on multiple processors I receive this assert after the first time step. Yes, the simulation starts and runs just fine through the first time step and dies during adaptivity.

```
Assertion `!node_touched_by_anyone[nodeid] || node_touched_by_me[nodeid]' failed.
[0] src/mesh/mesh_tools.C, line 1337, compiled nodate at notime
```

Is there a problem with doing the refinement up front BEFORE adding the equation systems? Is there information being lost by adding the equation system to a refined mesh that already has active children?

Stack traces from the two processes:
```
Stack frames: 16
0: 0   libmesh_dbg.0.dylib                 0x0000000111c07b32 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 82
1: 1   libmesh_dbg.0.dylib                 0x0000000111c0a130 libMesh::write_traceout() + 2656
2: 2   libmesh_dbg.0.dylib                 0x0000000111c01051 libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*) + 689
3: 3   libmesh_dbg.0.dylib                 0x00000001122bb3b3 void libMesh::MeshTools::libmesh_assert_valid_procids<libMesh::Node>(libMesh::MeshBase const&) + 5123
4: 4   libmesh_dbg.0.dylib                 0x0000000111abf3e0 libMesh::DofMap::distribute_local_dofs_var_major(unsigned int&, libMesh::MeshBase&) + 2544
5: 5   libmesh_dbg.0.dylib                 0x0000000111abbdc4 libMesh::DofMap::distribute_dofs(libMesh::MeshBase&) + 2644
6: 6   libmesh_dbg.0.dylib                 0x00000001126f895d libMesh::EquationSystems::reinit() + 3485
7: 7   libmoose-dbg.0.dylib                0x0000000110400ac4 FEProblem::meshChanged() + 196
8: 8   libmoose-dbg.0.dylib                0x00000001104009c4 FEProblem::adaptMesh() + 196
9: 9   libmoose-dbg.0.dylib                0x00000001107b7b70 Transient::incrementStepOrReject() + 96
10: 10  libmoose-dbg.0.dylib                0x00000001107b7a1d Transient::execute() + 125
11: 11  libmoose-dbg.0.dylib                0x0000000110607dce MooseApp::executeExecutioner() + 270
12: 12  libmoose-dbg.0.dylib                0x000000011060951c MooseApp::run() + 60
13: 13  moose_test-dbg                      0x000000010faab344 main + 292
14: 14  libdyld.dylib                       0x00007fff90f925c9 start + 1
15: 15  ???                                 0x0000000000000003 0x0 + 3
Stack frames: 18
0: 0   libmesh_dbg.0.dylib                 0x0000000111c07b32 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 82
1: 1   libmesh_dbg.0.dylib                 0x0000000111c0a130 libMesh::write_traceout() + 2656
2: 2   libmesh_dbg.0.dylib                 0x0000000111bb9ced libMesh::libmesh_terminate_handler() + 13
3: 3   libc++abi.dylib                     0x000000011728c628 _ZSt11__terminatePFvvE + 8
4: 4   libc++abi.dylib                     0x000000011728bc5b __cxa_throw + 171
5: 5   libmesh_dbg.0.dylib                 0x00000001122bb3fa void libMesh::MeshTools::libmesh_assert_valid_procids<libMesh::Node>(libMesh::MeshBase const&) + 5194
6: 6   libmesh_dbg.0.dylib                 0x0000000111abf3e0 libMesh::DofMap::distribute_local_dofs_var_major(unsigned int&, libMesh::MeshBase&) + 2544
7: 7   libmesh_dbg.0.dylib                 0x0000000111abbdc4 libMesh::DofMap::distribute_dofs(libMesh::MeshBase&) + 2644
8: 8   libmesh_dbg.0.dylib                 0x00000001126f895d libMesh::EquationSystems::reinit() + 3485
9: 9   libmoose-dbg.0.dylib                0x0000000110400ac4 FEProblem::meshChanged() + 196
10: 10  libmoose-dbg.0.dylib                0x00000001104009c4 FEProblem::adaptMesh() + 196
11: 11  libmoose-dbg.0.dylib                0x00000001107b7b70 Transient::incrementStepOrReject() + 96
12: 12  libmoose-dbg.0.dylib                0x00000001107b7a1d Transient::execute() + 125
13: 13  libmoose-dbg.0.dylib                0x0000000110607dce MooseApp::executeExecutioner() + 270
14: 14  libmoose-dbg.0.dylib                0x000000011060951c MooseApp::run() + 60
15: 15  moose_test-dbg                      0x000000010faab344 main + 292
16: 16  libdyld.dylib                       0x00007fff90f925c9 start + 1
17: 17  ???                                 0x0000000000000003 0x0 + 3
```

@friedmud, @roystgnr - Can you think of any call I'm missing?

**Comment by permcody** [comment]
Actually even that doesn't make sense. This PR passes the other 1000+ tests, many of which have uniform refinement turned on. Maybe there's a problem with the way we build our Refinement and Coarsening maps that's confusing the equation systems object.

**Comment by YaqiWang** [comment]
When is the mesh partitioned?

**Comment by roystgnr** [comment]
> [0] src/mesh/mesh_tools.C, line 1337, compiled nodate at notime

Now I want to put an assert in at line 80085 somewhere.

> Assertion `!node_touched_by_anyone[nodeid] || node_touched_by_me[nodeid]' failed.

What this means is that the mesh has a node in it which the current
processor believes it owns, which is connected to an element owned by
another processor, and which is not connected to an element owned by
this processor.

The problem should be independent of EquationSystems.  I'm not sure
what *could* lead to this situation, though.

**Comment by permcody** [comment]
How about doing several levels of refinement at once? I removed this logic since the equation systems don't exist at the point where I'm doing the uniform refinements.
https://github.com/idaholab/moose/blob/devel/framework/src/base/Adaptivity.C#L217

If that isn't it, I'll dig into https://github.com/idaholab/moose/blob/devel/framework/src/mesh/MooseMesh.C#L1378
Derek builds several independent meshes for each element type in the simulation to create "closest qp maps" for copying stateful materials when necessary. This is the only other thing that I can think of that might be screwing libMesh up.

**Comment by permcody** [comment]
OK - just to recheck for my own sanity. I changed the logic temporarily to force the uniform refinement in the original location (i.e. in FEProblem::initialSetup()) which is way late and after all systems have been setup. The test ran just fine without any assertions or other errors. Simply moving the refinement to an early spot is causing this error.

**Comment by roystgnr** [comment]
I believe doing several levels of refinement at once should be supported if-and-only-if there's no solutions being projected from the old to the new mesh.

Typically we *do* have solutions to be projected, so even when we have multiple refinement levels to do (e.g. when performing a fine restart from a very coarse solution) we typically do the refinement one level at a time.  But "start with a coarse mesh, then do uniformly_refine(N>1), then build systems on it and do further AMR" is such a common use case that I'd be astonished if it had regressed.

Unless the independent meshes are sharing nodes (which would have been too huge a bug to have gone uncaught to begin with), using multiple meshes shouldn't create any confusion about which processors own what.

How do I replicate the current failure case?  ./run_tests --something_or_other?

How small (n_elem) a failure case can you boil this down to?

**Comment by permcody** [comment]
Thanks - I don't think there's any issue with several levels of refinements or anything else you proposed. The MOOSE and application tests suites run fine with this change and in parallel. This issue with this one test is something much more stupid :smile: Hopefully @friedmud has an idea to shed some light on the situation.

**Comment by friedmud** [comment]
I'm stumped too.  I mean - what does this have to do with stateful material properties?!?  Or can you get this to happen without that now?

**Comment by permcody** [comment]
I believe you need stateful materials + adaptivity + uniform refinement to
trigger this problem.

**Comment by friedmud** [comment]
Looking at this now

**Comment by permcody** [comment]
Cool, if you need something let me know. I've been doing a lot of this work on one of the hpcbuild boxes. I just used X-forwarding and -start_in_debugger to trip the error. It's 100% reproducible on more than one processor.

One more thing. Just for fun I commented out the buildRefinementandCoarsening() maps call in FEProblem::initialSetup(). I STILL hit this error after the first timestep but before it would need those maps to do the material property copies.

**Comment by friedmud** [comment]
Ok - thanks for the heads up on that.  I'm going to go hardcore...
commenting out huge chunks of the system until I figure out where shit is
getting screwed up...

Derek

**Comment by permcody** [comment]
Other things of tried:

- Did one level of refinement at a time even before the equation system exists just to make sure doing all the levels at once wasn't a source of the problem - NO CHANGE
- Put "if false" in the first refinement step SetupMeshChanged::act() to disable early refinement and a corresponding "if true" in FEProblem::initialSetup() to restore the original order while leaving the rest of my PR in place. That indeed fixes it.

**Comment by permcody** [comment]
@roystgnr and @jwpeterson - @friedmud discovered that the issue had to do with our use of MeshBase::skip_partitioning(true) so I started messing around with that. We have this turned off because of #4532. It turns out that we are apparently calling that method "too late" but we have always called it rather late.

I hardcoded the skip partitioning call right after the uniform refinement calls and this test case ran. The problem is that we only want to disable partitioning for the very specific case spelled out in #4532. Now that we are doing these refinements first I don't have all of the information available to make that decision so I'm at a point where I'm trying to understand the specific issue of why this doesn't work to figure out the right place.

Current MOOSE workflow looks like this (simplified):
1. build or read mesh
2. prepare 
3. modify mesh (optional)
4. prepare again (if mesh was modified)
5. setup the rest of MOOSE
6. skip partitioning if applicable (#4532) 
6. start running simulation
7. build coarsening and refinement maps using mesh and mesh refinement utilities 
8. uniform refine

This PR moves step 8 to step 3.5 which breaks due to the skip partitioning in 6. I already mentioned that If I hard code 6 at 3.6 it fixes this case. What I don't understand is why this is issue. Clearly we are getting away with preparing the mesh in 2 and 4 without the skip partitioning flag now.

In a different attempt to fix the issue I attempted to add yet another mesh.prepare() call at 6.1 which did not change anything. I'm failing to understand why the partitioning appears to get hosed when moving around the refinement even after calling prepare yet again. Any insight would be helpful.

**Comment by permcody** [comment]
Wait! Even more interesting is that I successfully put step 6 at 4.1 and it worked. That leaves only step 5 in the middle so there's something else going on possibly with setting up the equation system. Still looking...

**Comment by jwpeterson** [comment]
On Thu, Feb 5, 2015 at 10:44 AM, Cody Permann <notifications@github.com>
wrote:

> @roystgnr <https://github.com/roystgnr> and @jwpeterson
> <https://github.com/jwpeterson> - @friedmud <https://github.com/friedmud>
> discovered that the issue had to do with our use of
> MeshBase::skip_partitioning(true) so I started messing around with that. We
> have this turned off because of #4532
> <https://github.com/idaholab/moose/issues/4532>. It turns out that we are
> apparently calling that method "too late" but we have always called it
> rather late.
>
> I hardcoded the skip partitioning call right after the uniform refinement
> calls and this test case ran. The problem is that we only want to disable
> partitioning for the very specific case spelled out in #4532
> <https://github.com/idaholab/moose/issues/4532>. Now that we are doing
> these refinements first I don't have all of the information available to
> make that decision so I'm at a point where I'm trying to understand the
> specific issue of why this doesn't work to figure out the right place.
>
> Current MOOSE workflow looks like this (simplified):
> 1. build or read mesh
> 2. prepare
> 3. modify mesh (optional)
> 4. prepare again (if mesh was modified)
> 5. setup the rest of MOOSE
> 6. skip partitioning if applicable (#4532
> <https://github.com/idaholab/moose/issues/4532>)
> 6. start running simulation
> 7. build coarsening and refinement maps using mesh and mesh refinement
> utilities
> 8. uniform refine
>
> This PR moves step 8 to step 3.5 which breaks due to the skip partitioning
> in 6. I already mentioned that If I hard code 6 at 3.6 it fixes this case.
> What I don't understand is why this is issue. Clearly we are getting away
> with preparing the mesh in 2 and 4 without the skip partitioning flag now.
>
In a different attempt to fix the issue I attempted to add yet another
> mesh.prepare() call at 6.1 which did not change anything.
>
I'm failing to understand why the partitioning appears to get hosed when
> moving around the refinement even after calling prepare yet again. Any
> insight would be helpful.
>
I'm not sure I understood all that, but what I think you are saying is that
uniform refining *before* preparing the mesh (while skipping partitioning)
breaks stateful material properties in MOOSE, while uniform refining after
later does not?

-- 
John

**Comment by permcody** [comment]
Don't look at it as "breaks stateful materials", instead I'm theorizing that it's actually breakage due to specific ordering of libMesh calls. I'm getting a libMesh assertion which I posted way up above before it gets to any of the material prolongation/restriction logic.

**Comment by friedmud** [comment]
Yes - this isn't actually a "stateful material properties" problem at all.
It's just that the codepath for stateful material properties is calling
`skip_partitioning()`.  This problem happens WITHOUT stateful material
properties if you call `skip_partitioning()`...

Like Cody says: there is some order dependence to uniform refine,
eq::setup(), skip_partitioning()

I suspect that basically no one else in the world ever uses
`skip_partitioning()` and some subtle side effect has crept in....

**Comment by permcody** [comment]
Whoa - I lied.  Turns out that due to multiple registration my disabling of the partition was happening earlier than I though.

I have found out that I have to disable partitioning before uniform refinement. The old workflow did that since the uniform refinement happened very late. It appears to still be a requirement. I don't have an easy solution to this problem unless I do something really hacky like disabling it and then turning it back on later (yuck). While I can snoop most pieces of information earlier if need be, we don't have a good way of snooping information about stateful material properties since they are added programmatically when Materials are added to the system.

**Comment by permcody** [comment]
I modified libMesh adaptivity ex2 to skip partitioning and to do the refinements before building the equation system and it ran just fine on multiple processors so there's still something in MOOSE that's causing this issue, not libMesh.

**Comment by friedmud** [comment]
I don't know - I wasn't able to trigger it in other MOOSE tests either... I
suspect that this is a real corner case. Can you try to make that libMesh
example as much like the MOOSE test as possible? Make it start with the
same mesh and do the same operations to it?

**Comment by permcody** [comment]
I got it! Takes four processors in the libMesh example. It won't fail on 2 or 3...

**Comment by permcody** [comment]
@roystgnr - I was able to create a test case by modifying a few lines in adaptivity_ex2 and running on 4 processors: https://github.com/permcody/libmesh/tree/adaptivity_ex2_assertion

If you simply comment out the `mesh.skip_partiitioning(true)` this example continues to work.

@jwpeterson and I noticed that if you print out the mesh info right before `equation_systems.reinit()` you'll see that the information is not the same when you toggle the skip_partitioning line. John conjectured that their may be some sort of contract operation happening in the partitioning step. Do you have any insight into the problem? 

With partitioning (working):
```
 Mesh Information:
  mesh_dimension()=2
  spatial_dimension()=3
  n_nodes()=46
    n_local_nodes()=14
  n_elem()=62
    n_local_elem()=10
    n_active_elem()=28
  n_subdomains()=2
  n_partitions()=4
  n_processors()=4
  n_threads()=1
  processor_id()=0
```

Without partitioning (not working):
```
 Mesh Information:
  mesh_dimension()=2
  spatial_dimension()=3
  n_nodes()=46
    n_local_nodes()=14
  n_elem()=62
    n_local_elem()=10
    n_active_elem()=28
  n_subdomains()=2
  n_partitions()=4
  n_processors()=4
  n_threads()=1
  processor_id()=0
```

**Comment by jwpeterson** [comment]
Specifically the `n_local_elem()` and `n_local_nodes()` numbers are different between the two cases.

**Comment by roystgnr** [comment]
I can replicate the adaptivity_ex2_assertion failure, thanks!  I'll see what I can figure out this afternoon.  The only guess that makes sense a priori is that skip_repartitioning() could *also* leave new child elements unpartitioned, but that shouldn't be; new children should be initialized with their parents' processor id in Elem::Elem().

**Comment by jwpeterson** [comment]
> new children should be initialized with their parents' processor id in Elem::Elem().

This was our first thought as well.  We tried hard-coding the setting of processor_id() in Elem::refine() to the parent's processor id, but it didn't seem to make any difference...

**Comment by YaqiWang** [comment]
I am just guessing: maybe the neighboring elements for the elements, which are the added elements with the refinement and also on the processor interface, are not set correctly? Because 2 or 3 processors does not fail, maybe it happens with a neighboring element connecting with an element through a node but not through a side?

**Comment by permcody** [comment]
Revisit #2412

**Comment by permcody** [comment]
@roystgnr - I'm not sure why you closed this branch. This is still a useful change that we'd like to merge into MOOSE, we just needed your commit in libMesh to make it possible. Was that a mistake?

**Comment by roystgnr** [comment]
*I'm* not sure *how* I closed this branch.  I was only trying to hit merge in the libMesh PR.

**Comment by permcody** [comment]
Github bug? That's odd

**Comment by permcody** [comment]
libMesh submodule update has been performed - This branch has been rebased so we should be able to merge once application patches have been applied

**Comment by friedmud** [comment]
Hmmm - does that mean it's still failing with the new libMesh?

**Comment by permcody** [comment]
Maybe, the tests passed on my system. Also one of the two tests passed so I
need to investigate more.

**Comment by permcody** [comment]
@friedmud - I don't believe this is a libMesh issue now. I compiled in debug mode and hit one of the new asserts that I threw into restrictStatefulProps():
```
Assertion `props().contains(child_elem)' failed
Child element pointer is not in the MaterialProps data structure
at /Users/permcj/projects/moose/framework/src/materials/MaterialPropertyStorage.C, line 224
```
Interestingly, the problem is fairly consistent. It _always_ runs on 2,4 or 8 procs, but _always_ fails on everything else in between. Since the mesh is created starting with a simple 2x2 grid with 3 uniform refinements, metis will produce a perfect partition for powers of 2. Every other combination of processors would result in different processors owning active children than the original parent partition. This is my current theory but even if I'm correct there isn't a good solution to the problem. Normally we "prepare" the mesh before setting up the system but we won't have the information available at that time as to whether we need to skip partitioning.

**Comment by friedmud** [comment]
Shouldn't matter what METIS does... because children should always be owned by the same processors as their parents... which is all that's necessary for the system to work.

I do agree that that looks like a problem in MOOSE now... just gonna have to debug...

**Comment by permcody** [comment]
Right, but that's exactly what's not happening here. If I have a 2x2 mesh that I uniform refine 3 times I end up a 16x16 mesh. I then call prepare() on that mesh and it gets partitioned based on the 16x16 mesh NOT the 2x2 mesh. We then add materials and figure out that we have a few stateful material properties declared so we skip partitioning at that point (after the initial partition was done on the refined mesh). Now when we try to coarsen.... boom!

**Comment by friedmud** [comment]
Well - there's no way around that then.  We need to set skip_partitioning() sooner...

**Comment by permcody** [comment]
We simply can't get that information earlier with the current API in MOOSE. At the point when the materials are added, so have all the variables and the equation systems (i.e. the current place where we skip and refine). So it's already "late".

We could try to skip then unskip...
Or I could just make a parameter in the [Problem] section for people that have these corner cases and need to control MOOSE (also not pretty - but temporary). That parameter could warn and even error if used incorrectly so it wouldn't be the end of the world.

At least it would allow us to move up the mesh refinement which could REALLY help some of the larger runs that Yaqi has that have expensive projections. Might also turn up the heat to fix this whole system.

**Comment by YaqiWang** [comment]
Not sure if I am understanding this correctly so let me rephrase it: The level of the elements on the uniformly refined mesh is not zero like the elements read in from the file or generated. This means that those elements could be deleted by coarsening, which will cause trouble in the stateful material system.

If this is the case, can we tell libmesh that these elements need to be treated as the initial elements and cannot be coarsened? For example, we can set their level to zero.

**Comment by permcody** [comment]
No, that's not the issue and the proper fix isn't to change the level of the elements. That would remove the ability to do coarsening at all for this case.

**Comment by YaqiWang** [comment]
If a user want to coarsening beyond the uniformly refined mesh, then he should not use this earlier uniform refinement at all. Instead, he can perform uniform refinement at initial as we did before.

**Comment by permcody** [comment]
While you can uniformly refine the mesh later if you want. We've always supported uniform refinement as a means of creating a suitable mesh for coarsening.

**Comment by YaqiWang** [comment]
I then think your proposed parameter is a good solution. Maybe it just needs to be in Mesh block. `delay_uniform_refinement` defaults to true. In case that it can be set to false, MOOSE issues a warning.

**Comment by friedmud** [comment]
@permcody We actually already have an option in `FileMesh` called `skip_partitioning`... go ahead and just make that a global option for all Mesh types and lets move on from this for now.  Just make it throw an error if you're trying to use stateful properties without setting "skip_partitioning".

(BTW: Do we want to invert that option?  Just make it `partition = false` to turn off partitioning?  `dont_undo_skip_partitioning` ?!? :imp:  Other option is to use the `partitioner` parameter we have now... and set it to `none` to turn off partitioning.  I actually kind of like that).

Anyway - we're trying to solve a problem that is somewhat "transient" as far as we hope to remove this restriction from stateful properties pretty soon anyway...

**Comment by permcody** [comment]
The parameter solution is fine. It's even more restricted than your case.
It takes a mesh with uniform refinement + stateful material props +
adaptivity to hit this issue. Very few users will ever run into this. I
like the idea of inverting the parameter.
